File: 250__Count_Univalue_Subtrees.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def countUnivalSubtrees(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.result = 0
        def recurse(node, parent):
            if not node:
                return True
            left = recurse(node.left, node.val)
            right = recurse(node.right, node.val)
            if left and right:
                self.result += 1
            return left and right and node.val == parent
        recurse(root, None)
        return self.result

File: Java/250__Count_Univalue_Subtrees.java
public class Solution {
    int count = 0;

    public int countUnivalSubtrees(TreeNode root) {
        if (root == null) return 0;
        helper(root);
        return count;
    }

    private boolean helper(TreeNode node) {
        if (node.left == null && node.right == null) {
            count++;
            return true;
        }
        boolean check = true;
        if (node.left != null)
            check = helper(node.left) && check && node.left.val == node.val;
        if (node.right != null)
            check = helper(node.right) && check && node.right.val == node.val;
        if (!check)
            return false;
        count++;
        return true;
    }
}
